File: gsabino365/lib/data/models/user_model.dart
class UserModel {
  final String? id;
  final String? name;
  final String? role;
  final String? email;
  final String? image;
  final String? status;
  final bool? verified;
  final String? subRole;
  final String? phoneNumber;
  final String? phone;
  final String? languagesSpoken;
  final String? preferredAttorney;
  final String? preferredBailBondsman;
  final String? licensedStatesToPractice;
  final String? barAssociationNumber;
  final String? lawFirmName;
  final String? officeName;
  final String? datePassedTheBar;
  final String? medicalLicenseNumber;
  final String? specialization;
  final String? specialty;
  final String? companyName;
  final String? businessAddress;
  final String? badgeNumber;
  final String? assignedNumber;
  final String? departmentOrPrecinct;
  final String? didCarNumberChange;
  final String? newCarNumber;
  final String? licenseNumber;
  final String? shortHexId;
  final String? country;
  final String? gender;
  final String? dateOfBirth;
  final String? profilePicture;
  final bool? isOnboardingCompleted;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  UserModel({
    this.id,
    this.name,
    this.role,
    this.email,
    this.image,
    this.status,
    this.verified,
    this.subRole,
    this.phoneNumber,
    this.phone,
    this.languagesSpoken,
    this.preferredAttorney,
    this.preferredBailBondsman,
    this.licensedStatesToPractice,
    this.barAssociationNumber,
    this.lawFirmName,
    this.officeName,
    this.datePassedTheBar,
    this.medicalLicenseNumber,
    this.specialization,
    this.specialty,
    this.companyName,
    this.businessAddress,
    this.badgeNumber,
    this.assignedNumber,
    this.departmentOrPrecinct,
    this.didCarNumberChange,
    this.newCarNumber,
    this.licenseNumber,
    this.shortHexId,
    this.country,
    this.gender,
    this.dateOfBirth,
    this.profilePicture,
    this.isOnboardingCompleted,
    this.createdAt,
    this.updatedAt,
  });

  factory UserModel.fromJson(Map<String, dynamic> json) {
    return UserModel(
      id: json['id'] ?? json['_id'],
      name: json['name'],
      role: json['role']?.toString().toUpperCase(),
      email: json['email'],
      image: json['image'] ?? json['profilePicture'],
      status: json['status'],
      verified: json['verified'],
      subRole: json['subRole'],
      phoneNumber: json['phoneNumber'] ?? json['phone'],
      phone: json['phone'] ?? json['phoneNumber'],
      languagesSpoken: json['languagesSpoken'],
      preferredAttorney: json['preferredAttorney'],
      preferredBailBondsman: json['preferredBailBondsman'],
      licensedStatesToPractice: json['licensedStatesToPractice'],
      barAssociationNumber: json['barAssociationNumber'],
      lawFirmName: json['lawFirmName'],
      officeName: json['officeName'],
      datePassedTheBar: json['datePassedTheBar'],
      medicalLicenseNumber: json['medicalLicenseNumber'],
      specialization: json['specialization'] ?? json['specialty'],
      specialty: json['specialty'] ?? json['specialization'],
      companyName: json['companyName'],
      businessAddress: json['businessAddress'],
      badgeNumber: json['badgeNumber'],
      assignedNumber: json['assignedNumber'],
      departmentOrPrecinct: json['departmentOrPrecinct'],
      didCarNumberChange: json['didCarNumberChange'],
      newCarNumber: json['newCarNumber'],
      licenseNumber: json['licenseNumber'],
      shortHexId: json['shortHexId']?.toString() ??
          (json['id'] != null || json['_id'] != null
              ? ((json['id'] ?? json['_id']).toString().length >= 8
                  ? (json['id'] ?? json['_id'])
                      .toString()
                      .substring(
                          (json['id'] ?? json['_id']).toString().length - 8)
                      .toUpperCase()
                  : (json['id'] ?? json['_id']).toString().toUpperCase())
              : null),
      country: json['country'],
      gender: json['gender'],
      dateOfBirth: json['dateOfBirth'],
      profilePicture: json['profilePicture'] ?? json['image'],
      isOnboardingCompleted: json['isOnboardingCompleted'],
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt'])
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.tryParse(json['updatedAt'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'role': role,
      'email': email,
      'image': image,
      'status': status,
      'verified': verified,
      'subRole': subRole,
      'phoneNumber': phoneNumber,
      'phone': phone,
      'languagesSpoken': languagesSpoken,
      'preferredAttorney': preferredAttorney,
      'preferredBailBondsman': preferredBailBondsman,
      'licensedStatesToPractice': licensedStatesToPractice,
      'barAssociationNumber': barAssociationNumber,
      'lawFirmName': lawFirmName,
      'officeName': officeName,
      'datePassedTheBar': datePassedTheBar,
      'medicalLicenseNumber': medicalLicenseNumber,
      'specialization': specialization,
      'companyName': companyName,
      'businessAddress': businessAddress,
      'badgeNumber': badgeNumber,
      'assignedNumber': assignedNumber,
      'departmentOrPrecinct': departmentOrPrecinct,
      'didCarNumberChange': didCarNumberChange,
      'newCarNumber': newCarNumber,
      'licenseNumber': licenseNumber,
      'shortHexId': shortHexId,
      'country': country,
      'gender': gender,
      'dateOfBirth': dateOfBirth,
      'profilePicture': profilePicture,
      'isOnboardingCompleted': isOnboardingCompleted,
      'createdAt': createdAt?.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
    };
  }
}

File: apps/backend/src/app/modules/user/user.model.ts
import bcrypt from 'bcrypt';
import { StatusCodes } from 'http-status-codes';
import { model, Schema } from 'mongoose';
import config from '../../../config';
import { USER_ROLES } from '../../../enums/user';
import ApiError from '../../../errors/ApiError';
import { IUser, UserModal } from './user.interface';

const userSchema = new Schema<IUser, UserModal>(
  {
    name: {
      type: String,
      required: true,
    },
    role: {
      type: String,
      enum: Object.values(USER_ROLES),
      required: true,
    },
    email: {
      type: String,
      required: true,
      lowercase: true,
    },
    password: {
      type: String,
      required: true,
      select: 0,
      minlength: 8,
    },
    image: {
      type: String,
      default: 'https://i.ibb.co/z5YHLV9/profile.png',
    },
    status: {
      type: String,
      enum: ['active', 'inactive', 'delete'],
      default: 'active',
    },
    verified: {
      type: Boolean,
      default: false,
    },
      authentication: {
      type: {
        isResetPassword: {
          type: Boolean,
          default: false,
        },
        oneTimeCode: {
          type: Number,
          default: null,
        },
        expireAt: {
          type: Date,
          default: null,
        },
      },
      select: 0,
    },
    subRole: { type: String },
    phoneNumber: { type: String },
    languagesSpoken: { type: String },
    preferredAttorney: { type: String, default: '' },
    preferredBailBondsman: { type: String, default: '' },
    licensedStatesToPractice: { type: String },
    barAssociationNumber: { type: String },
    lawFirmName: { type: String },
    officeName: { type: String },
    datePassedTheBar: { type: String },
    medicalLicenseNumber: { type: String },
    specialization: { type: String },
    companyName: { type: String },
    businessAddress: { type: String },
    badgeNumber: { type: String },
    assignedNumber: { type: String },
    departmentOrPrecinct: { type: String },
    didCarNumberChange: { type: String },
    newCarNumber: { type: String },
    licenseNumber: { type: String },
  },
  {
    timestamps: true,
    toJSON: {
      virtuals: true,
      transform: (_doc, ret: Record<string, unknown>) => {
        ret.id = ret._id;
        return ret;
      },
    },
    toObject: { virtuals: true },
  }
);

userSchema.virtual('shortHexId').get(function (this: { _id?: { toString(): string } }) {
  const idStr = this._id ? this._id.toString() : '';
  return idStr.length >= 8 ? idStr.slice(-8).toUpperCase() : idStr.toUpperCase();
});

userSchema.index({ email: 1, role: 1 }, { unique: true });

//exist user check
userSchema.statics.isExistUserById = async (id: string) => {
  const isExist = await User.findById(id);
  return isExist;
};

userSchema.statics.isExistUserByEmail = async (email: string) => {
  const isExist = await User.findOne({ email });
  return isExist;
};

userSchema.statics.isExistUserByEmailAndRole = async (email: string, role: string) => {
  const isExist = await User.findOne({ email, role });
  return isExist;
};

//is match password
userSchema.statics.isMatchPassword = async (
  password: string,
  hashPassword: string
): Promise<boolean> => {
  return await bcrypt.compare(password, hashPassword);
};

//check user
userSchema.pre('save', async function (next) {
  if (this.isNew) {
    const isExist = await User.findOne({ email: this.email, role: this.role });
    if (isExist) {
      throw new ApiError(StatusCodes.BAD_REQUEST, 'Email and Role combination already exists!');
    }
  }

  //password hash
  if (this.isModified('password')) {
    this.password = await bcrypt.hash(
      this.password,
      Number(config.bcrypt_salt_rounds)
    );
  }
  next();
});

export const User = model<IUser, UserModal>('User', userSchema);
